### ISSUE TITLE
Check inferred `this` parameter against interface requirement `this: void`

### ISSUE THREAD

**Issue by jtbandes** [microsoft/TypeScript#44513]
# Bug Report

### 🔎 Search Terms

interface method requirement this: void

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.2#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIC2EYAFugCYAUZwAzgFzIBu6wFAlEyAK6EBG0ANy4AvrgQAbOHTrIAogA84hAA6SUwNRuLg5GLHgKLkAXmQBmEQWJlKVDjmRQSPKCGS06AOkVDk4qJAA)

### 💻 Code

```ts
interface Foo {
  method(this: void): number;
}
class Example implements Foo {
  x = 3;
  method() { return this.x; } // no error
}
```

### 🙁 Actual behavior

No error

### 🙂 Expected behavior

Error because the method implementation's implicit `this: Example` is not compatible with the interface requirement `this: void`.

The [`@typescript-eslint/unbound-method` lint rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/unbound-method.md) uses `this: void` as an annotation indicating the method is allowed to be used without binding `this`. However, there seems to be no checking from the TS compiler as to whether an inferred `this` type in a method implementation matches an interface requirement. Thus the lint rule can help you add `this: void` to the interface if you intend for functions to be usable when not bound... but neither the linter nor compiler currently help prevent runtime errors resulting from implementations with mismatched `this` types.

**Comment by sandersn** [comment]
As I recall, the implicit `this` for methods is still `this: any` not `this: this`, which was an intentional decision for backward compatibility. It might be time to revisit this decision. I'll investigate during the 4.4 beta.

**Comment by fatcerberus** [comment]
IIRC, I think I recall @RyanCavanaugh saying the reason for that is because `this: this` being the default implicitly makes all methods generic, which was a performance concern.

**Comment by fatcerberus** [comment]
Found it: https://github.com/microsoft/TypeScript/issues/35451#issuecomment-565116770

**Comment by sandersn** [comment]
Yes, the only difference between `this: Example` and `this: this` in the above example is which type you get. The checker creates `this<Example>` at the same time it creates `Example`.

The next question is what breaks when methods have `this: this` by default.

**Comment by thatsmydoing** [comment]
If performance concerns are an issue, would it make sense to add this as an option like `strictMethodThis`?
